### PR TITLE
feat: add MissingJsDoc and MissingTypeRef diagnostics

### PR DESCRIPTION
**Pull request by dsherret**
It's not perfect, but probably good enough for the first pass. We should infer return types like `void` when there's no return statement.

### CODE REVIEW

File: src/diagnostics.rs [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    }
+  }
+}
+
+struct DiagnosticDocNodeVisitor<'a> {
+  diagnostics: &'a mut DiagnosticsCollector,
+}
+
+impl<'a> DiagnosticDocNodeVisitor<'a> {
+  pub fn visit_doc_nodes(&mut self, doc_nodes: &[DocNode]) {
```
Comment (bartlomieju) [inline src/diagnostics.rs:115]: So it's pretty much a lint rule? 😅

File: src/tests.rs [hunk 2]
```diff
@@ -749,10 +786,14 @@ async fn exports_imported_earlier_default() {
...
-    .parse_with_reexports(&specifier)
-    .unwrap();
+  let entries = DocParser::new(
+    &graph,
+    analyzer.as_capturing_parser(),
+    DocParserOptions::default(),
+  )
+  .unwrap()
+  .parse_with_reexports(&specifier)
+  .unwrap();
```
Comment (bartlomieju) [inline src/tests.rs:796]: Something for a follow up - add a helper that will set it all up and hide `unwrap`s

**Review by bartlomieju — APPROVED** [review]
Fantastic!

File: src/tests.rs (hunk 2 above)
Comment (dsherret) [reply src/tests.rs:796]: Yeah, I think a lot of these tests could be moved to the specs folder

File: src/diagnostics.rs (hunk 1 above)
Comment (dsherret) [reply src/diagnostics.rs:110]: Yes, but much smarter than something like https://github.com/denoland/deno_lint/pull/972

Comment (bartlomieju) [reply src/diagnostics.rs:110]: Should we close that issue because of this feature?

Comment (dsherret) [reply src/diagnostics.rs:110]: Yeah, I'll have the CLI feature PR close it.

**Pull request merged**